generic ephemeral volume: add metrics

pohly opened the pull request:
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
As discussed during the production readiness review, a metric for the
PVC create operations is useful. The "ephemeral_volume" workqueue
metrics were already added in the initial implementation.

#### Special notes for your reviewer:

The new code follows the example set by the endpoints controller.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
added `ephemeral_volume_controller_create[_failures]_total` counters to kube-controller-manager metrics
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/pohly/enhancements/blob/master/keps/sig-storage/1698-generic-ephemeral-volumes/README.md#monitoring-requirements
```

/sig storage

pohly committed:
generic ephemeral volume: add metrics

As discussed during the production readiness review, a metric for the
PVC create operations is useful. The "ephemeral_volume" workqueue
metrics were already added in the initial implementation.

The new code follows the example set by the endpoints controller.

pohly commented:
/retest

jsafrane commented:
/assign @gnufied

gnufied commented on pkg/controller/volume/ephemeral/metrics/metrics.go:
```diff
@@ -0,0 +1,53 @@
+			StabilityLevel: metrics.ALPHA,
+		},
+		[]string{"reason"},
+	)
```
Can you provide an example of how metrics appear in prometheus?

pohly replied:
My tentative update of the KEP has such examples: https://github.com/kubernetes/enhancements/pull/2513/files

Is that what you were looking for or shall I add it here, for example to the commit message?

gnufied replied:
ah nvm. I also see this captured in release-notes.

gnufied commented:
Before we merge this change - do we know this metric can be extracted from api-server logs or api-server somehow? I would think that api-server logs should contain some information about incoming PVC creation requests (may not be as metric directly).

pohly commented:
> I would think that api-server logs should contain some information about incoming PVC creation requests (may not be as metric directly).

The ask from the production readiness review was explicitly to add metrics. Extracting the same information via log analysis might work, but it's simply not the same thing.

gnufied commented:
cc @kubernetes/sig-instrumentation-members  @logicalhan

gnufied commented:
> The ask from the production readiness review was explicitly to add metrics. Extracting the same information via log analysis might work, but it's simply not the same thing.

I understand. I am bit concerned about creating approximately 20-25 new timeseries metric because reason is being used a a dimension.

pohly commented:
> I am bit concerned about creating approximately 20-25 new timeseries metric because reason is being used a a dimension.

I have no strong opinion either way. I can implement whatever is considered appropriate for this feature. We can also reduce the status to just success/failure, status just seemed a bit more flexible. If we use a boolean, what are a good name and values for such a dimension?

brancz commented on pkg/controller/volume/ephemeral/controller.go:
```diff
@@ -279,6 +283,14 @@ func (ec *ephemeralController) handleVolume(pod *v1.Pod, vol v1.Volume) error {
+			reason = "Unknown"
+		}
+	}
+	ephemeralvolumemetrics.EphemeralVolumeCreate.WithLabelValues(reason).Inc()
```
Is `reason` practically bound here? What kind of values are we expecting here for this to be valuable?

brancz commented on pkg/controller/volume/ephemeral/controller.go:
```diff
@@ -279,6 +283,14 @@ func (ec *ephemeralController) handleVolume(pod *v1.Pod, vol v1.Volume) error {
 		Spec: ephemeral.VolumeClaimTemplate.Spec,
 	}
 	_, err = ec.kubeClient.CoreV1().PersistentVolumeClaims(pod.Namespace).Create(context.TODO(), pvc, metav1.CreateOptions{})
+	reason := ""
```
Having `reason` be an empty string is going to make querying this metric awkward. I think we probably want a label that distinguishes between success and failure, or even have separate metrics for those.

pohly replied:
I'm fine with a label `success` that has values `yes` and `no`. Is that okay or are there better names? I am not sure what is used elsewhere.

brancz replied:
The best practice is to have two metrics one for failures and one for all. Then you can naturally query `failures / total` and understand the percentage. The number of series you produce is identical, so this doesn't produce any more load or anything on the metrics system than what we already have here.

pohly replied:
I've changed to to count `ephemeral_volume_controller_create_total` and `ephemeral_volume_controller_create_failures`. Is that okay?

Note that my updated tests initially failed because `GetCounterMetricValue` didn't work for a Counter. I'm including the fix for that (= adding the missing Metric interface implementation).

brancz replied:
Yep, that works!

pohly committed:
k8s.io/component-base: support GetCounterMetricValue for Counter

GetCounterMetricValue has an unchecked conversion to metrics.Metric,
something that didn't work for a *Counter because it didn't implement
that interface.

pohly committed:
generic ephemeral volume: simpler metrics

A CounterVector with status as label may create unnecessary overhead
and using the success case with the empty label value wasn't
easy. It's better to have two seperate counters, one for total number
of calls and one for failed calls.

fedebongio commented:
/remove-sig api-machinery

pohly commented:
/retest

brancz commented:
cc @logicalhan for the component base part (which looks good to me just would like him to double check), otherwise lgtm

ehashman commented:
/assign @brancz 
for SIG Instrumentation

/triage accepted

gnufied commented:
/lgtm

pohly commented:
@gnufied I had to rebase to get rid of the modified BUILD file. Can you lgtm again?

@brancz Can I get an lgtm from you, too? logicalhan is out.

pohly commented:
/remove-sig api-machinery

brancz commented:
/lgtm
/approve

pohly commented:
/retest

brancz commented on pkg/controller/volume/ephemeral/metrics/metrics.go:
```diff
@@ -0,0 +1,58 @@
+	EphemeralVolumeCreateFailures = metrics.NewCounter(
+		&metrics.CounterOpts{
+			Subsystem:      EphemeralVolumeSubsystem,
+			Name:           "create_failures",
```
I apologize for not catching this earlier. The guidelines are to always end counters with `_total` so this one needs that as well.

pohly replied:
Done, please take another look.

brancz reviewed (commented):
Just a tiny naming change to align with guidelines

brancz commented:
/lgtm
/approve

pohly commented:
/retest

brancz commented:
/lgtm
